### PR TITLE
Switch to using more expressive layout descriptors instead of numeric layout names

### PR DESCRIPTION
**Pull request by jbobba**
This is a relatively big PR but most changes are cosmetic. Feel free to review a sub-set of files/changes.

### CODE REVIEW

File: src/ngraph/runtime/cpu/mkldnn_utils.cpp [hunk 1]
```diff
@@ -209,18 +209,211 @@ const std::string& runtime::cpu::mkldnn_utils::get_mkldnn_format_string(memory::
...
+        mkldnn_memory_desc_t named_md;
+        // Could throw an exception if named `format` is not compatible with `md.dims`
+        error::wrap_c_api(
+            mkldnn_memory_desc_init(&named_md, blk.ndims, blk.dims, blk.data_type, format), "");
+        return memory::desc(named_md);
+    };
+
+    auto compare_named_md = [&](const mkldnn_memory_desc_t& blk,
+                                const mkldnn_memory_format_t format,
+                                const memory::desc& out_md) {
```
Comment (pruthviIntel) [inline src/ngraph/runtime/cpu/mkldnn_utils.cpp]:
``` /ec/fm/disks/nrv_algo_home01/pthoreho/ngraph/src/ngraph/runtime/cpu/mkldnn_utils.cpp:379:53: warning: declaration shadows a local
      variable [-Wshadow]
                                const memory::desc& out_md) {
                                                    ^
/ec/fm/disks/nrv_algo_home01/pthoreho/ngraph/src/ngraph/runtime/cpu/mkldnn_utils.cpp:367:10: note: previous declaration is here
    auto out_md = memory::desc(md);
```
I get this warning here, need to change the variable name in lambda function

File: src/ngraph/runtime/cpu/pass/cpu_layout.cpp [hunk 2]
```diff
@@ -339,134 +362,126 @@ namespace ngraph
                         }
                     }
                     convolution_forward::primitive_desc prim_desc(*fwd_desc, cpu_engine);
-                    prim_input_formats.push_back(static_cast<memory::format>(
-                        prim_desc.src_primitive_desc().desc().data.format));
+                    auto src_prim_desc = prim_desc.src_primitive_desc().desc();
```
Comment (pruthviIntel) [inline src/ngraph/runtime/cpu/pass/cpu_layout.cpp]: unused variable.

```
/ec/fm/disks/nrv_algo_home01/pthoreho/ngraph/src/ngraph/runtime/cpu/pass/cpu_layout.cpp:370:26: warning: unused variable
      'src_prim_desc' [-Wunused-variable]
                    auto src_prim_desc = prim_desc.src_primitive_desc().desc();
```

File: src/ngraph/runtime/cpu/pass/cpu_layout.cpp [hunk 3]
```diff
@@ -1332,101 +1376,79 @@ namespace ngraph
...
                 void CPULayout::LAYOUT_DECL(ngraph::op::Concat)
                 {
-                    if (runtime::cpu::mkldnn_utils::use_mkldnn_kernel(node.get()))
+                    if (mkldnn_utils::use_mkldnn_kernel(node.get()))
                     {
                         auto concat = static_cast<const ngraph::op::Concat*>(node.get());
-                        auto input0_layout =
-                            runtime::cpu::mkldnn_utils::get_input_mkldnn_format(node.get(), 0);
-                        size_t num_inputs = node->get_input_size();
+                        auto input0_md = mkldnn_utils::get_input_mkldnn_md(node.get(), 0);
```
Comment (pruthviIntel) [inline src/ngraph/runtime/cpu/pass/cpu_layout.cpp]: unused variable 

```
/ec/fm/disks/nrv_algo_home01/pthoreho/ngraph/src/ngraph/runtime/cpu/pass/cpu_layout.cpp:1408:30: warning: unused variable
      'input0_md' [-Wunused-variable]
                        auto input0_md = mkldnn_utils::get_input_mkldnn_md(node.get(), 0);
```

**Comment by pruthviIntel** [comment]
review is WIP. I will try to review everything soon.

File: src/ngraph/runtime/cpu/pass/cpu_layout.cpp [hunk 4]
```diff
@@ -157,35 +174,41 @@ void runtime::cpu::pass::CPULayout::set_default_layouts(
...
-            auto layout =
-                std::make_shared<ngraph::runtime::cpu::LayoutDescriptor>(*tv, native_axis_order);
-            layout->set_mkldnn_format(runtime::cpu::mkldnn_utils::CreateNativeDataFormat(*cpu_tvl));
-            auto new_node = std::shared_ptr<Node>(
-                new runtime::cpu::op::ConvertLayout(output.get_node(), output.get_index(), layout));
-            new_args.push_back(new_node);
-            if (use_replace)
+            auto native_md =
+                mkldnn_utils::create_blocked_mkldnn_md(shape, cpu_tvl->get_strides(), et);
+            if (!mkldnn_utils::compare_mkldnn_mds(cpu_tvl->get_mkldnn_md(), native_md))
```
Comment (pruthviIntel) [inline src/ngraph/runtime/cpu/pass/cpu_layout.cpp]: little confused here, if the node already have mkldnn layout and if we again create mkldnn md based on tv->strides . isn't it both are same?. and here when we meant native_md is it NCHW? (framework format)

Comment (jbobba) [reply src/ngraph/runtime/cpu/pass/cpu_layout.cpp]: cpu_tvl->get_strides() is always row-major. get_mkldnn_md can be whatever the layout pass has on that tensor.

File: src/ngraph/runtime/cpu/cpu_external_function.cpp [hunk 5]
```diff
@@ -980,17 +978,19 @@ using namespace ngraph::runtime;
 void runtime::cpu::CPU_ExternalFunction::propagate_in_place_input(
     ngraph::descriptor::Output* output, std::string input_name)
 {
-    auto it = output;
-    auto propagate_further = false;
-    do
+    std::deque<ngraph::descriptor::Output*> stack;
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/cpu_external_function.cpp]: This really could've been a separate PR  :)

File: src/ngraph/runtime/cpu/cpu_layout_descriptor.cpp [hunk 6]
```diff
@@ -18,63 +18,46 @@
...
-            const AxisVector LayoutDescriptor::CHWNAxisOrder{1, 2, 3, 0};
-
-            AxisVector LayoutDescriptor::create_native_axis_order(size_t rank)
-            {
-                AxisVector native_axis_order(rank);
-                std::iota(native_axis_order.begin(), native_axis_order.end(), 0);
-                return native_axis_order;
-            }
+            const mkldnn::memory::desc
+                LayoutDescriptor::DummyDesc(mkldnn::memory::dims(TENSOR_MAX_DIMS),
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/cpu_layout_descriptor.cpp]: `UninitializedDesc` ?

File: src/ngraph/runtime/cpu/cpu_emitter.cpp [hunk 7]
```diff
@@ -1760,13 +1702,34 @@ namespace ngraph
             void CPU_Emitter::EMITTER_DECL(ngraph::op::Reshape)
             {
                 auto reshape = static_cast<const ngraph::op::Reshape*>(node);
-                if (!reshape->get_is_transpose() && out[0].get_name() == args[0].get_name())
+                auto can_skip_reshape = [&]() {
+                    if (!reshape->get_is_transpose())
+                    {
+                        return true;
+                    }
+                    auto annotation = reshape->get_op_annotations();
```
[the remark this answers is not in the document]
Comment (jbobba) [reply src/ngraph/runtime/cpu/cpu_emitter.cpp]: There are cases where cpu layout eliminates reshapes that are transposes as well.

File: src/ngraph/runtime/cpu/builder/reshape.cpp [hunk 8]
```diff
@@ -42,134 +42,95 @@ namespace ngraph
...
-                    Shape input_shape(input_axis_order.size());
-                    for (size_t idx = 0; idx < input_axis_order.size(); idx++)
+                auto can_skip_reshape = [&]() {
+                    if (!reshape->get_is_transpose())
                     {
-                        input_shape[idx] = args[0].get_shape()[input_axis_order[idx]];
+                        return true;
                     }
+                    auto annotation = reshape->get_op_annotations();
+                    if (annotation && annotation->get_in_place_oi_pairs().size() > 0)
```
[the remark this answers is not in the document]
Comment (jbobba) [reply src/ngraph/runtime/cpu/builder/reshape.cpp]: In line 51, the presence of an OI pair is an indication that the op isn't destructive and hence gives license to the emitter to skip the kernel.

File: src/ngraph/runtime/cpu/mkldnn_utils.cpp [hunk 9]
```diff
@@ -209,18 +209,211 @@ const std::string& runtime::cpu::mkldnn_utils::get_mkldnn_format_string(memory::
...
+        }
+        catch (const mkldnn::error&)
+        {
+            // Cannot create the named descriptor compatible with `in` desc
+            return false;
+        }
+        return false;
+    };
+
+#define CANONICALIZE_MD(X)                                                                         \
```
[the remark this answers is not in the document]
Comment (jbobba) [reply src/ngraph/runtime/cpu/mkldnn_utils.cpp]: We can pull this out into its own method if there are more places we need to use it.

File: src/ngraph/runtime/cpu/pass/cpu_layout.cpp [hunk 10]
```diff
@@ -999,139 +987,214 @@ namespace ngraph
...
+                                       ->get_tensor_view_layout();
+                        auto cpu_tvl = dynamic_cast<runtime::cpu::LayoutDescriptor*>(tvl.get());
+                        if (cpu_tvl && cpu_tvl->is_mkldnn_layout())
+                        {
+                            // Rotate MKLDNN memory descriptor
+                            auto input_md = mkldnn_utils::get_input_mkldnn_md(node.get(), 0);
+                            auto output_md = mkldnn_utils::rotate_blocked_md(input_md, axis_order);
+                            set_output_layouts(node, {output_md});
+                            auto op_annotations = reshape->get_op_annotations();
+                            if (op_annotations)
```
[the remark this answers is not in the document]
Comment (jbobba) [reply src/ngraph/runtime/cpu/pass/cpu_layout.cpp]: No good reason other than paranoia.

File: src/ngraph/runtime/cpu/cpu_external_function.cpp (hunk 5 above)
Comment (jbobba) [reply src/ngraph/runtime/cpu/cpu_external_function.cpp]: Agree. But its there now and you cant unreview it.

File: src/ngraph/runtime/cpu/cpu_layout_descriptor.cpp (hunk 6 above)
Comment (jbobba) [reply src/ngraph/runtime/cpu/cpu_layout_descriptor.cpp]: I dont care either way. I can change it if you feel strongly about it or you can sneak it in another PR.

File: src/ngraph/runtime/cpu/cpu_layout_descriptor.hpp [hunk 11]
```diff
@@ -37,36 +36,36 @@ namespace ngraph
...
                 size_t get_index_offset(const std::vector<size_t>& indices) override;
 
-                const Strides& get_strides() const override { return strides; }
+                const Strides& get_strides() const override { return m_strides; }
+                void set_strides(Strides& strides) { m_strides = strides; }
                 bool operator==(const TensorViewLayout& other) const override;
 
-                void set_mkldnn_format(const mkldnn::memory::format& format)
+                const mkldnn::memory::desc& get_mkldnn_md() const { return m_mkldnn_md; }
+                void set_mkldnn_md(const mkldnn::memory::desc md) { m_mkldnn_md = md; }
```
Comment (nbpatel) [inline src/ngraph/runtime/cpu/cpu_layout_descriptor.hpp]: where is this getting called?

Comment (pruthviIntel) [reply src/ngraph/runtime/cpu/cpu_layout_descriptor.hpp]: in layout pass.

Comment (nbpatel) [reply src/ngraph/runtime/cpu/cpu_layout_descriptor.hpp]: ah ok, got it. github requires us to manually press the "Load Diff" if the file is too big.  Hence my Command + F couldn't locate it

**Pull request merged**